samples/core/tutorials/estimators/linear.ipynb

```diff
@@ -1399,6 +1422,211 @@
+        "\n",
+        "model_l1.train(train_inpf)\n",
+        "\n",
+        "results = model_l1.evaluate(test_inpf)\n",
+        "clear_output()\n",
+        "for key in sorted(results):\n",
+        "  print('%s: %0.2f' % (key, results[key]))"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "ofmPL212JIy2",
+        "colab_type": "code",
+        "colab": {
+          "autoexec": {
+            "startup": false,
+            "wait_interval": 0
+          }
+        }
+      },
+      "cell_type": "code",
+      "source": [
+        "model_l2 = tf.estimator.LinearClassifier(\n",
+        "    feature_columns=base_columns + crossed_columns,\n",
+        "    optimizer=tf.train.FtrlOptimizer(\n",
+        "        learning_rate=0.1,\n",
+        "        l1_regularization_strength=0.0,\n",
+        "        l2_regularization_strength=10.0))\n",
+        "\n",
+        "model_l2.train(train_inpf)\n",
+        "\n",
+        "results = model_l2.evaluate(test_inpf)\n",
+        "clear_output()\n",
+        "for key in sorted(results):\n",
+        "  print('%s: %0.2f' % (key, results[key]))"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "Lp1Rfy_k4e7w",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "These regularized models don't perform very differently base model. Let look ar the models' weight distributions to better see the effect of the regularization:"
+      ]
+    },
+    {
+      "metadata": {
+        "id": "Wb6093N04XlS",
+        "colab_type": "code",
+        "colab": {
+          "autoexec": {
+            "startup": false,
+            "wait_interval": 0
+          }
+        }
+      },
+      "cell_type": "code",
+      "source": [
+        "def get_flat_weights(model):\n",
+        "  weight_names = [\n",
+        "      name for name in model.get_variable_names()\n",
+        "      if \"linear_model\" in name and \"Ftrl\" not in name]\n",
+        "\n",
+        "  weight_values = [model.get_variable_value(name) for name in weight_names]\n",
+        "\n",
+        "  weights_flat = np.concatenate([item.flatten() for item in weight_values], axis=0)\n",
+        "\n",
+        "  return weights_flat\n",
+        "\n",
+        "weights_flat = get_flat_weights(model)\n",
+        "weights_flat_l1 = get_flat_weights(model_l1)\n",
+        "weights_flat_l2 = get_flat_weights(model_l2)"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "metadata": {
+        "id": "GskJmtfmL0p-",
+        "colab_type": "text"
+      },
+      "cell_type": "markdown",
+      "source": [
+        "The models have many zero-vlaued weights  caused by unused hash bins (There are many more hash bins than categories in some columns). We will mask these weights when viewing the weight distributions:"
```

Comment:
Copy:

The models have many zero-valued weights caused by unused hash bins (there are many more hash bins than categories in some columns). We can mask these weights when viewing the weight distributions:

Reply:
Done